airflow/models.py

```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
         qry = session.query(DR).filter(DR.dag_id == dag_id)
         if run_id:
             qry = qry.filter(DR.run_id == run_id)
+        if execution_date:
+            qry = qry.filter(DR.execution_date == execution_date)
         if state:
             qry = qry.filter(DR.state == state)
         if external_trigger:
             qry = qry.filter(DR.external_trigger == external_trigger)
-        dr = qry.all()
+        dr = qry.order_by(DR.execution_date).all()
 
         return dr
 
+    @provide_session
+    def get_task_instances(self, state=None, session=None):
+        """
+        Returns the task instances for this dag run
+        """
+        TI = TaskInstance
+        tis = session.query(TI).filter(
+            TI.dag_id == self.dag_id,
+            TI.execution_date == self.execution_date,
+        )
+        if state:
+            if isinstance(state, six.string_types):
+                tis = tis.filter(TI.state == state)
+            else:
+                # this is required to deal with NULL values
+                if None in state:
+                    tis = tis.filter(or_(
+                        TI.state.in_(state),
+                        TI.state.is_(None))
+                    )
+                else:
+                    tis = tis.filter(TI.state.in_(state))
+
+        return tis.all()
+
+    @provide_session
+    def get_task_instance(self, task_id, session=None):
+        """
+        Returns the task instance specified by task_id for this dag run
+        :param task_id: the task id
+        """
+
+        TI = TaskInstance
+        ti = session.query(TI).filter(
+            TI.dag_id == self.dag_id,
+            TI.execution_date == self.execution_date,
+            TI.task_id == task_id
+        ).one()
+
+        return ti
+
+    def get_dag(self):
+        """
+        Returns the Dag associated with this DagRun
+        :param session: database session
+        :return: DAG
+        """
+        if not self.dag:
+            raise AirflowException("{}.dag needs to be set".format(self))
```

Comment:
nit: this is really minor but I don't think it will be clear from this error message that `XXX.dag` is referring to the DAG attribute of this DagRun. I think it will look something like `<DagRun foo bar>.dag`. Maybe saying `The DAG for {} needs to be set` would be more clear?